src/Service/Resolvers/ArrayPoolWriter.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+
+/// <summary>
+/// A helper to write to pooled arrays.
+/// </summary>
+internal sealed class ArrayPoolWriter : IBufferWriter<byte>, IDisposable
+{
+    private const int _initialBufferSize = 512;
+    private byte[] _buffer;
+    private int _capacity;
+    private int _start;
+    private bool _disposed;
+
+    public ArrayPoolWriter()
+    {
+        _buffer = ArrayPool<byte>.Shared.Rent(_initialBufferSize);
+        _capacity = _buffer.Length;
+        _start = 0;
+    }
+    
+    public ReadOnlyMemory<byte> GetWrittenMemory() 
+        => _buffer.AsMemory().Slice(0, _start);
+
+    public ReadOnlySpan<byte> GetWrittenSpan() 
+        => _buffer.AsSpan().Slice(0, _start);
+
+    public void Advance(int count)
+    {
+        _start += count;
+        _capacity -= count;
+    }
+
+    public Memory<byte> GetMemory(int sizeHint = 0)
+    {
```

Comment:
Shouldn't GetMemory/GetSpan functions verify its not disposed before accessing the buffer.AsMemory?